### ISSUE TITLE
Enhance distributed-training-coordinator agent with pre-RDMA sync, coordinator chunking, and validator integration

### ISSUE THREAD

**Issue by akaszubski** [akaszubski/autonomous-dev#283]
## Summary

Enhance the existing distributed-training-coordinator agent with pre-RDMA synchronization validation, coordinator-level dataset chunking, and integration with HardwareCalibrator, WorkerConsistencyValidator, and DistributedTrainingValidator. This update adds 5 new workflow phases to prevent OOM errors, ensure codebase consistency across workers, and validate distributed training configurations before execution.

## CRITICAL CORRECTIONS (Measured Performance Data)

**Key findings from actual distributed training runs:**

1. **Equal Machine Performance (Not 65/35 Split)**
   - Previous assumption: Heterogeneous performance requiring 65/35 workload split
   - **Reality**: Both machines run at ~0.85 ex/s for the scoring pipeline (equal performance)
   - **Impact**: Workload distribution logic should assume equal split, not dynamic ratio
   - **Action**: Update Phase 2.5 Hardware Calibration to measure actual throughput, not assume ratios

2. **macOS Process Priority (`nice=5` Ineffective)**
   - Current approach: Using `nice=5` for process prioritization
   - **Problem**: macOS ignores `nice` values - POSIX nice() is stubbed on Darwin
   - **Solution**: Must use macOS QoS (Quality of Service) class API instead
   - **Action**: Replace `nice=5` with QoS API calls (e.g., `pthread_set_qos_class_self_np()` or Python `os.setpriority()` with PRIO_DARWIN_THREAD)

3. **Pipeline Overhead-Bound (Not Compute-Bound)**
   - Previous measurement: 5.1x speedup reported
   - **Reality**: The 5.1x ratio was anomalous - actual pipeline is overhead-bound
   - **Impact**: Scaling efficiency is limited by coordination overhead, not compute
   - **Action**: Focus optimization on reducing coordinator overhead (chunking, async I/O, pipeline parallelism)

**References:**
- Measured on: macOS Darwin with MLX distributed training
- Date: January 2025
- Pipeline: Scoring pipeline with ~0.85 examples/second per machine

---

### Files to Modify

- **`/Users/andrewkaszubski/Dev/autonomous-dev/plugins/autonomous-dev/agents/distributed-training-coordinator.md`**: Update agent workflow with new phases

### Changes Required

1. **Add Phase 1.5: Pre-RDMA Sync Validation** (after Phase 1: Setup Analysis)
   ```markdown
   ### Phase 1.5: Pre-RDMA Sync Validation
   
   Before distributed training begins, validate codebase consistency across all workers:
   
   1. Run pre-RDMA sync script: `~/Dev/sync-dev.sh`
   2. Validate all workers have identical:
      - Plugin versions
      - Python dependencies
      - MLX library versions
      - Training scripts
   3. Block distributed launch if sync fails
   4. Report sync status with clear error messages
   
   **Rationale**: RDMA distributed training requires identical code across all workers. Missing or mismatched dependencies cause cryptic failures mid-training.
   ```

2. **Add Phase 2.5: Hardware Calibration** (after Phase 2: Scaling Strategy)
   ```markdown
   ### Phase 2.5: Hardware Calibration
   
   Measure actual hardware throughput and distribute workload accordingly:
   
   1. Invoke HardwareCalibrator (from Issue #280):
      - Measure per-GPU throughput (samples/sec) - **MEASURED, NOT ASSUMED**
      - **CRITICAL**: Both machines show ~0.85 ex/s (equal performance, not 65/35)
      - Detect memory limits per GPU
      - Identify stragglers (GPUs >20% slower than mean)
   2. Adjust per-GPU batch size based on calibration
   3. Plan gradient accumulation to balance stragglers
   4. Store calibration results in training plan JSON
   5. **macOS Process Priority**:
      - **DO NOT USE**: `nice=5` (macOS ignores POSIX nice values)
      - **USE INSTEAD**: QoS API via `os.setpriority(os.PRIO_DARWIN_THREAD, 0, qos_class)`
      - QoS classes: QOS_CLASS_USER_INITIATED (high priority), QOS_CLASS_UTILITY (background)
   
   **Integration Point**: `plugins/autonomous-dev/lib/agent_invoker.py` - Add HardwareCalibrator to AGENT_CONFIGS
   
   **Performance Reality**:
   - Pipeline is **overhead-bound**, not compute-bound
   - 5.1x speedup was anomalous - real scaling limited by coordinator overhead
   - Focus on reducing coordination costs (chunking, async I/O, pipeline parallelism)
   ```

3. **Add Phase 3.5: Worker Consistency Validation** (after Phase 3: Distributed Configuration)
   ```markdown
   ### Phase 3.5: Worker Consistency Validation
   
   Validate critical files and configurations across all workers:
   
   1. Invoke WorkerConsistencyValidator (from Issue #281):
      - Validate training script SHA256 hash
      - Validate model checkpoint existence
      - Validate random seed synchronization
      - Validate MLX environment variables
   2. Report any mismatches with specific file paths
   3. Block mlx.launch if validation fails
   4. Provide fix commands (rsync, scp, export)
   
   **Integration Point**: `plugins/autonomous-dev/lib/workflow_coordinator.py` - Add to invoke_parallel_validators()
   ```

4. **Add Phase 4.5: Coordinator-Level Chunking** (after Phase 4: Verification Plan)
   ```markdown
   ### Phase 4.5: Coordinator-Level Chunking
   
   Prevent OOM errors on large datasets by chunking at the coordinator level:
   
   1. Check dataset size:
      - If dataset has >50K examples, activate chunking
      - Set COORDINATOR_CHUNK_SIZE = 50000
   2. Process dataset in chunks:
      ```python
      for chunk_start in range(0, len(dataset), COORDINATOR_CHUNK_SIZE):
          chunk = dataset[chunk_start:chunk_start + COORDINATOR_CHUNK_SIZE]
          # Distribute chunk to workers
          # Train on chunk
          # Collect gradients
          gc.collect()
          mx.clear_cache()  # Clear MLX cache
      ```
   3. Report chunk progress (Chunk 1/5, Chunk 2/5, ...)
   4. Add chunking config to training plan JSON
   
   **Rationale**: Datasets >50K examples can exceed coordinator memory. Chunking prevents OOM while maintaining distributed efficiency.
   
   **Overhead Consideration**: Chunking adds coordination overhead but is necessary for memory safety. Given pipeline is already overhead-bound, chunking cost is acceptable trade-off.
   ```

5. **Add Phase 5: Pre-Flight Checklist** (new final phase)
   ```markdown
   ### Phase 5: Pre-Flight Checklist
   
   Run comprehensive validation before invoking mlx.launch:
   
   1. Invoke DistributedTrainingValidator (from Issue #282):
      - ✓ RDMA devices detected (ibstat)
      - ✓ NCCL environment variables set correctly
      - ✓ MLX distributed training vars configured
      - ✓ Workers are reachable via network
      - ✓ Training script exists on all workers
      - ✓ Model checkpoint accessible
      - ✓ Dataset chunking configured (if >50K examples)
      - ✓ FlashRecovery checkpoint storage accessible
   2. Report validation results with clear pass/fail status
   3. Block mlx.launch if any check fails
   4. Provide fix commands for each failure
   
   **Integration Point**: `plugins/autonomous-dev/lib/workflow_coordinator.py` - Add to invoke_parallel_validators()
   ```

6. **Update Output Format**
   - Add new fields to training plan JSON:
     - `pre_rdma_sync_status`: { "synced": true, "workers": 4, "timestamp": "..." }
     - `hardware_calibration`: { "per_gpu_throughput": [850, 850, 848, 852], "equal_performance": true, "overhead_bound": true }
     - `worker_consistency`: { "script_hash": "abc123", "checkpoint_validated": true }
     - `chunking_config`: { "enabled": true, "chunk_size": 50000, "total_chunks": 5 }
     - `pre_flight_validation`: { "passed": true, "failed_checks": [] }
     - `macos_qos_config`: { "qos_class": "QOS_CLASS_USER_INITIATED", "nice_ignored": true }

7. **Backward Compatibility**
   - All new phases are optional enhancements
   - Existing Phase 1-4 workflows continue to work
   - New phases activate automatically when validators are available
   - Graceful degradation if validators not installed

### 1. Fresh Install with New Validators
- **Given**: New autonomous-dev installation with Issues #280, #281, #282 validators installed
- **When**: distributed-training-coordinator agent is invoked for 100K example dataset
- **Then**:
  - Phase 1.5: Runs ~/Dev/sync-dev.sh and validates worker consistency
  - Phase 2.5: HardwareCalibrator measures per-GPU throughput (~0.85 ex/s per machine)
  - Phase 3.5: WorkerConsistencyValidator checks training script hash
  - Phase 4.5: Coordinator chunking activates (2 chunks of 50K)
  - Phase 5: All 8 pre-flight checks pass
  - mlx.launch executes with correct configuration

### 2. Update to Existing Agent (Backward Compatibility)
- **Given**: Existing distributed-training-coordinator agent in use
- **When**: Agent is updated with new phases
- **Then**:
  - Existing Phase 1-4 workflows continue to work
  - New phases activate only if validators are available
  - No breaking changes to existing training plans
  - JSON output format is backward compatible (new fields optional)

### 3. Large Dataset Chunking (>50K Examples)
- **Given**: Dataset with 200K training examples
- **When**: distributed-training-coordinator agent creates training plan
- **Then**:
  - Phase 4.5 detects dataset size >50K
  - Coordinator chunking activates automatically
  - 4 chunks of 50K examples each
  - gc.collect() + mx.clear_cache() after each chunk
  - Chunking config added to training plan JSON

### 4. Validator Integration (Pre-Flight Checks)
- **Given**: Distributed training setup with 4 GPUs on 2 nodes
- **When**: Phase 5 pre-flight checklist runs
- **Then**:
  - DistributedTrainingValidator checks all 8 validation points
  - RDMA detection: ibstat confirms InfiniBand devices
  - NCCL vars: NCCL_IB_DISABLE=0, NCCL_NET_GDR_LEVEL=5
  - MLX vars: MLX_DISTRIBUTED_ENABLED=1, MLX_DISTRIBUTED_RANK set per worker
  - Worker reachability: ping/ssh to all nodes succeeds
  - Training script: SHA256 hash matches across workers
  - Model checkpoint: Exists and accessible from all workers
  - Chunking: Configured correctly for dataset size
  - FlashRecovery: Storage path accessible (gs://, s3://, nfs://)

### 5. Validation Failure with Fix Commands
- **Given**: Worker consistency validation fails (mismatched training script)
- **When**: Phase 3.5 runs WorkerConsistencyValidator
- **Then**:
  - Validation fails with clear error: "Training script mismatch on worker-2"
  - Provides fix command: `rsync -avz train.py worker-2:/path/to/train.py`
  - Blocks mlx.launch until issue is resolved
  - User runs fix command and re-invokes agent
  - Validation passes, training proceeds

### 6. Pre-RDMA Sync Failure
- **Given**: ~/Dev/sync-dev.sh fails due to network issue
- **When**: Phase 1.5 runs pre-RDMA sync validation
- **Then**:
  - Sync fails with clear error: "Worker-3 unreachable during sync"
  - Provides fix command: `ssh worker-3 'cd ~/Dev && git pull'`
  - Blocks distributed training until sync succeeds
  - User fixes network issue and re-invokes agent
  - Sync succeeds, training proceeds

### 7. macOS Process Priority (QoS API)
- **Given**: Training on macOS with distributed coordinator
- **When**: Phase 2.5 attempts to set process priority
- **Then**:
  - Detects macOS platform (platform.system() == "Darwin")
  - **DOES NOT USE**: `os.nice(5)` (would be ignored)
  - **USES INSTEAD**: `os.setpriority(os.PRIO_DARWIN_THREAD, 0, QOS_CLASS_USER_INITIATED)`
  - Logs: "macOS QoS set to USER_INITIATED (high priority)"
  - Training plan JSON includes: `"macos_qos_config": {"qos_class": "QOS_CLASS_USER_INITIATED"}`

### Updates to Issue #274 Components
- [ ] `distributed-training-coordinator.md` updated with 5 new phases (1.5, 2.5, 3.5, 4.5, 5)
- [ ] Backward compatibility preserved (existing Phase 1-4 workflows still work)
- [ ] Integration with HardwareCalibrator (Issue #280) in Phase 2.5
- [ ] Integration with WorkerConsistencyValidator (Issue #281) in Phase 3.5
- [ ] Integration with DistributedTrainingValidator (Issue #282) in Phase 5
- [ ] Training plan JSON output format updated with new fields
- [ ] **macOS QoS API implemented** (replaces ineffective `nice=5`)
- [ ] **Equal performance assumption** (both machines ~0.85 ex/s)
- [ ] **Overhead-bound pipeline documented** (not compute-bound)

### Pre-RDMA Sync (Phase 1.5)
- [ ] Runs ~/Dev/sync-dev.sh before distributed work begins
- [ ] Validates codebase consistency across all workers
- [ ] Blocks distributed launch if sync fails
- [ ] Provides clear error messages with fix commands

### Hardware Calibration (Phase 2.5)
- [ ] HardwareCalibrator measures per-GPU throughput (not assumes ratios)
- [ ] **Validates equal performance** (~0.85 ex/s per machine)
- [ ] Adjusts per-GPU batch size based on calibration results
- [ ] Identifies stragglers (GPUs >20% slower than mean)
- [ ] Stores calibration results in training plan JSON
- [ ] **macOS: Uses QoS API**, not `nice=5`
- [ ] **Documents overhead-bound nature** of pipeline

### Worker Consistency (Phase 3.5)
- [ ] WorkerConsistencyValidator checks training script SHA256 hash
- [ ] Validates model checkpoint existence on all workers
- [ ] Validates random seed synchronization
- [ ] Blocks mlx.launch if validation fails
- [ ] Provides fix commands (rsync, scp, export)

### Coordinator Chunking (Phase 4.5)
- [ ] Activates automatically for datasets >50K examples
- [ ] COORDINATOR_CHUNK_SIZE = 50000 prevents OOM
- [ ] gc.collect() + mx.clear_cache() after each chunk
- [ ] Reports chunk progress (Chunk 1/5, Chunk 2/5, ...)
- [ ] Chunking config added to training plan JSON

### Pre-Flight Validation (Phase 5)
- [ ] All 8 pre-flight checks run before mlx.launch
- [ ] RDMA devices detected (ibstat)
- [ ] NCCL environment variables validated
- [ ] MLX distributed training vars validated
- [ ] Workers reachable via network (ping/ssh)
- [ ] Training script SHA256 hash validated
- [ ] Model checkpoint accessible from all workers
- [ ] Dataset chunking configured correctly
- [ ] FlashRecovery storage accessible
- [ ] Clear error messages with fix commands for each failure
- [ ] Blocks mlx.launch if any check fails

### Validation
- [ ] Clear pass/fail status for each validation phase
- [ ] Specific file paths in error messages
- [ ] Fix commands provided for each failure type
- [ ] Graceful degradation if validators not installed

### Testing
- [ ] Integration test in `tests/integration/test_training_agents.py`:
  - Test all 5 new phases execute correctly
  - Test validator integration (HardwareCalibrator, WorkerConsistencyValidator, DistributedTrainingValidator)
  - Test coordinator chunking activates for large datasets
  - Test pre-RDMA sync validation
  - Test pre-flight checklist runs all 8 checks
  - **Test macOS QoS API** (not `nice=5`)
  - **Test equal performance calibration** (~0.85 ex/s)
- [ ] Smoke test in `tests/regression/smoke/test_training_agents_load.py`:
  - Verify distributed-training-coordinator agent loads without errors
  - Verify agent invocation completes successfully
  - Verify training plan JSON output format is valid

---

**Related Issues**:
- Part of #274 (Training best practices agents and skills)
- Depends on #280 (HardwareCalibrator)
- Depends on #281 (WorkerConsistencyValidator)
- Depends on #282 (DistributedTrainingValidator)
- Builds on #279 (MLX environment variables)

**Environment Requirements**:
- MLX distributed training library
- RDMA/InfiniBand hardware (optional, graceful fallback to TCP/IP)
- OpenShift AI 2.19+ (optional, for RDMA support)
- Python 3.9+
- Issues #280, #281, #282 validators installed
- **macOS**: Requires QoS API support (Darwin kernel)

**Source of Truth**:
- MLX distributed training docs: https://ml-explore.github.io/mlx/build/html/usage/distributed.html
- NCCL RDMA configuration: https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/env.html
- FlashRecovery paper: "FlashRecovery: Enabling Fast Checkpoint Recovery in Large-Scale Training" (2024)
- macOS QoS API: https://developer.apple.com/documentation/dispatch/dispatchqos
- Verified on OpenShift AI 2.19 with InfiniBand networking (January 2025)
- **Performance data**: Measured on macOS Darwin with MLX, ~0.85 ex/s per machine (January 2025)
